WS-554 - Extract Top Bar into a new component

louisearchibald opened the pull request:
Resolves JIRA: [WS-554](https://bbc.atlassian.net/browse/WS-554)

Summary
======
Takes the Top Bar Onward Journey component code that was previously intertwined with our existing ScrollablePromo component as part of an Optimizely experiment, and extracts it out into a brand new component of it's own, bringing it in line with our component standards. 

The component will be rendered on Article pages only at mobile breakpoints.

Code changes
======
- Creates a `TopBarOJs` component based on our existing ScrollablePromo legacy component, but converting to modern React & Typescript used elsewhere in our codebase.
- Adds emotion css styling scoped to each part of the component.
- Creates a `index.stories.tsx` file so the component can be viewed in our Storybook component library.
- Implements click and view tracking functionality.
- Adds tests to check component rendering and logic, and event tracking.
- Adds a README and metadata.json.
- Adds an `isLive()` check around the component for now, so it will not render on the live environment.
- Calls the component from the Navigation Header and passes component props to all relevant places.
- Updates bundle sizes.


Developer Checklist
======

- [ ] **UX**
  - [ ] UX Criteria met (visual UX & screenreader UX)
- [ ] **Accessibility**
    - [ ] Accessibility Acceptance Criteria met
    - [ ] Accessibility swarm completed
    - [ ] Component Health updated
    - [ ] P1 accessibility bugs resolved 
    - [ ] P2/P3 accessibility bugs planned (if not resolved)
- [ ] **Security**
    - [ ] Security issues addressed
    - [ ] Threat Model updated
- [ ] **Documentation**
    - [ ] Docs updated (runbook, READMEs)
- [ ] **[Testing](#testing)** 
    - [ ] Feature tested on relevant environments
- [ ] **Comms** 
    - [ ] Relevant parties notified of changes


Testing
======

- [ ] Manual Testing required?
  - [ ] Local (`Ready-For-Test, Local`)
  - [ ] Test  (`Ready-For-Test, Test`)
  - [ ] Preview (`Ready-For-Test, Preview`)
  - [ ] Live (`Ready-For-Test, Live`)
- [ ] Manual Testing complete?
  - [ ] Local
  - [ ] Test
  - [ ] Preview
  - [ ] Live


## Additional Testing Steps
1. Can be viewed locally by checking out the branch and navigating to `http://localhost:7080/russian/articles/c61q94n3rm3o`. NOTE: The component will not render at Desktop size, so the browser window will have to be made smaller in order for the component to appear.


Useful Links
======
- [Storybook link to component example](https://ws-554-extract-top-bar-into-new-compo--5d28eb3fe163f6002046d6fa.chromatic.com/iframe.html?globals=&args=&id=components-topbarojs--example&viewMode=story)
- [Storybook link to component example with a Live item](https://ws-554-extract-top-bar-into-new-compo--5d28eb3fe163f6002046d6fa.chromatic.com/iframe.html?globals=&args=&id=components-topbarojs--oj-top-bar-top-stories-with-live-label&viewMode=story)

louisearchibald committed:
Merge branch 'latest' of ssh://github.com/bbc/simorgh into latest
Merge branch 'latest' of ssh://github.com/bbc/simorgh into latest
:wq:wq# the commit.

louisearchibald committed:
Merge branch 'latest' of ssh://github.com/bbc/simorgh into latest

louisearchibald committed:
Merge branch 'latest' of ssh://github.com/bbc/simorgh into latest

louisearchibald committed:
Merge branch 'latest' of ssh://github.com/bbc/simorgh into latest

louisearchibald committed:
Merge branch 'latest' of ssh://github.com/bbc/simorgh into latest

louisearchibald committed:
Merge branch 'latest' of ssh://github.com/bbc/simorgh into latest

louisearchibald committed:
Merge branch 'latest' of ssh://github.com/bbc/simorgh into latest

louisearchibald committed:
Merge branch 'latest' of ssh://github.com/bbc/simorgh into latest

louisearchibald committed:
Merge branch 'latest' of ssh://github.com/bbc/simorgh into latest

louisearchibald committed:
Merge branch 'latest' of ssh://github.com/bbc/simorgh into latest

louisearchibald committed:
Merge branch 'latest' of ssh://github.com/bbc/simorgh into latest

louisearchibald committed:
Merge branch 'latest' of ssh://github.com/bbc/simorgh into latest

louisearchibald committed:
take copy of scrollable promo component

louisearchibald committed:
import path

louisearchibald committed:
update component name and export

louisearchibald committed:
remove experiment logic

louisearchibald committed:
remove unused imports and styled components

louisearchibald committed:
cleanup experiment logic inside PromoList

louisearchibald committed:
update imports

louisearchibald committed:
cleanup experiment logic inside Promo

louisearchibald committed:
remove switch case and refactor

louisearchibald committed:
remove timestamp logic and unused imports

louisearchibald committed:
call component from article page

louisearchibald committed:
yay the component is on the page with data

louisearchibald committed:
Merge branch 'latest' into WS-554-extract-top-bar-into-new-component

louisearchibald committed:
Merge branch 'WS-554-extract-top-bar-into-new-component' of ssh://github.com/bbc/simorgh into WS-554-extract-top-bar-into-new-component

louisearchibald committed:
imports

louisearchibald committed:
update storybook to see component locally

louisearchibald committed:
rename file to tsx

louisearchibald committed:
define props for component

louisearchibald committed:
change Promo index to tsx

louisearchibald committed:
update PromoList to tsx

louisearchibald committed:
start typescriptifying PromoList

louisearchibald committed:
further update to PromoList

louisearchibald committed:
converted styles for PromoList

louisearchibald committed:
style conversions added for Promo

louisearchibald committed:
apply styles and remove service and script references

louisearchibald committed:
remove Promo styled components

louisearchibald committed:
delete unused imports

louisearchibald committed:
clean up references to dir and remove styled components

louisearchibald committed:
change div to section and pass a11y ids

louisearchibald committed:
delete tests relating to mostRead

louisearchibald committed:
delete tests relating to mostRead

louisearchibald committed:
remove styled components

louisearchibald committed:
incorrect syntax

louisearchibald committed:
incorrect syntax and update timestamp test

louisearchibald committed:
delete unused style

louisearchibald committed:
update to props

louisearchibald committed:
fix type error

louisearchibald committed:
remove Grid component from return

louisearchibald committed:
amend media queries so component does not show at desktop

louisearchibald committed:
delete import

louisearchibald committed:
fix section tag

louisearchibald committed:
revert type change

louisearchibald committed:
remove mostRead from fixture data

louisearchibald committed:
amend promo and promoList

louisearchibald committed:
resolve type error

louisearchibald committed:
comment removal

louisearchibald committed:
add readme and metadata file for stories file update

louisearchibald committed:
update main test file

louisearchibald committed:
update TopBar tests to use component, tests failing still to be fixed

louisearchibald committed:
Merge branch 'latest' into WS-554-extract-top-bar-into-new-component

louisearchibald committed:
correct media queries

louisearchibald committed:
Merge branch 'WS-554-extract-top-bar-into-new-component' of ssh://github.com/bbc/simorgh into WS-554-extract-top-bar-into-new-component

louisearchibald committed:
fix padding value

louisearchibald committed:
should be a ul not section tag

louisearchibald committed:
final fixes to styling

louisearchibald committed:
fix a11y bug by making component visible with high contrast

louisearchibald committed:
make title strong element

louisearchibald committed:
fix focus on link

louisearchibald committed:
fix test failures

louisearchibald committed:
remove console log

louisearchibald committed:
further test file update

louisearchibald committed:
Merge branch 'latest' into WS-554-extract-top-bar-into-new-component

louisearchibald committed:
Merge branch 'WS-554-extract-top-bar-into-new-component' of ssh://github.com/bbc/simorgh into WS-554-extract-top-bar-into-new-component

louisearchibald committed:
fix type error inside ArticlePage

louisearchibald committed:
import order

louisearchibald committed:
Merge branch 'latest' into WS-554-extract-top-bar-into-new-component

louisearchibald committed:
suppress ts error

louisearchibald committed:
Merge branch 'WS-554-extract-top-bar-into-new-component' of ssh://github.com/bbc/simorgh into WS-554-extract-top-bar-into-new-component

louisearchibald committed:
increaase bundle size

louisearchibald committed:
fix for a11y test

louisearchibald committed:
amend component return

louisearchibald committed:
small change

louisearchibald committed:
trying to fix a11y violations

louisearchibald committed:
Merge branch 'latest' into WS-554-extract-top-bar-into-new-component

louisearchibald committed:
Merge branch 'WS-554-extract-top-bar-into-new-component' of ssh://github.com/bbc/simorgh into WS-554-extract-top-bar-into-new-component

louisearchibald committed:
Merge branch 'latest' into WS-554-extract-top-bar-into-new-component

louisearchibald committed:
revert previous change

louisearchibald committed:
Merge branch 'WS-554-extract-top-bar-into-new-component' of ssh://github.com/bbc/simorgh into WS-554-extract-top-bar-into-new-component

louisearchibald committed:
import

louisearchibald committed:
change to aria-label

louisearchibald committed:
delete tests and update snapshots

amoore108 commented on src/app/components/TopBarOJs/index.styles.ts:
```diff
@@ -0,0 +1,76 @@
+          marginInlineStart: '0',
+          padding: '0 16px',
+          margin: '0 -0.8rem',
+        },
```
Think we can remove the `&&` here.

The `padding` could use `rem` instead of `px`.

amoore108 commented on src/app/components/TopBarOJs/index.styles.ts:
```diff
@@ -0,0 +1,76 @@
+        padding: `0 ${spacings.DOUBLE}rem ${spacings.DOUBLE}rem`,
+        margin: '0 -0.2rem',
+      },
+      [mq.GROUP_3_MIN_WIDTH]: {
+        margin: '0 -0.8rem',
+      },
```
May be able to remove this to fix the margin on tablet sizes.

amoore108 commented:
Looks good!

I think the actual component needs to go in this file rather than in the ArticlePage: https://github.com/bbc/simorgh/blob/aae34f9201b02783bed4480d25b7b1c0c3c6ae60/src/app/legacy/containers/Navigation/index.canonical.jsx#L74-L81

It resides in the actual `Header` component. 

You may just be able to do a drop-in replacement of the `ScrollablePromo` component that is there already with your new one though.

The `blocks` data comes from here: https://github.com/bbc/simorgh/blob/1ec084b954fe18e55d2470c6cbb5521668d9d742/src/app/components/PageLayoutWrapper/index.tsx#L88-L103 which is still checking for experiment values though, so that will need removed and always be set to pass down `topStories`.

May require some prop renaming, as the `propsForOJExperiment` doesn't make too much sense anymore: https://github.com/bbc/simorgh/blob/1ec084b954fe18e55d2470c6cbb5521668d9d742/src/app/components/PageLayoutWrapper/index.tsx#L250

louisearchibald committed:
delete unused fixtures

louisearchibald committed:
rename stories to tsx and update file

louisearchibald committed:
import order

louisearchibald committed:
fix margin

pvaliani commented on src/app/components/TopBarOJs/PromoList/index.tsx:
```diff
@@ -0,0 +1,57 @@
+  return (
+    <ul
+      css={scrollablePromoStyles}
+      role="list"
```
is the role of `list` redundant here since its an unordered list?

louisearchibald replied:
Good shout, I was following what the component was doing from before and it has `role=list` here but it probably is redundant if the list if unordered, so I will remove it.

pvaliani reviewed (approved):
Looks good! great job

louisearchibald committed:
remove role=list

louisearchibald committed:
remove commented code

louisearchibald committed:
call component from correct location

louisearchibald committed:
remove experiment code

louisearchibald committed:
remove import

louisearchibald committed:
remove experimentVariant props and directly pass topStories to blocks

louisearchibald committed:
unused import

louisearchibald committed:
add test to check component is not rendered when isLive is true

louisearchibald committed:
Merge branch 'latest' into WS-554-extract-top-bar-into-new-component

louisearchibald committed:
add click and view tracking tests

louisearchibald committed:
Merge branch 'WS-554-extract-top-bar-into-new-component' of ssh://github.com/bbc/simorgh into WS-554-extract-top-bar-into-new-component

louisearchibald committed:
update bundle size

louisearchibald committed:
update event tracking tests

louisearchibald committed:
update README

louisearchibald committed:
linting

louisearchibald committed:
remove import

louisearchibald committed:
remove duplicate test file

louisearchibald committed:
delete experiment test

louisearchibald committed:
unused import

louisearchibald committed:
remove unused fixture blocks from fixture helper file

amoore108 committed:
Fixes

- Fix margin-left styling causing whitespace
- Fix eventTracking type
- Pass topStories data directly to component rather than through an object

amoore108 committed:
Merge branch 'latest' into WS-554-extract-top-bar-into-new-component

amoore108 committed:
Fix background colour - Remove duplicate view tracker

LilyL0u commented:
Looks good. I think we should probably have a think about what we want to do with the fact that there are now two very similar 'Top Tori' sections on the page. It doesn't appear to cause any a11y errors 👏 , but maybe a bit odd? Without the continue reading button in place as well, it isn't that bad as although they are repeated, we don't see them at the same time. But if we have the continue reading button on the page too, then we may be able to see both Top Tori sections at the same time... We shall see.

<img width="891" height="147" alt="Screenshot 2025-08-12 at 12 10 58" src="https://github.com/user-attachments/assets/bf8b4304-38ff-47dc-8ca3-6adeec87686b" />


<img width="915" height="165" alt="Screenshot 2025-08-12 at 12 11 06" src="https://github.com/user-attachments/assets/cca8d4a9-5b2e-41fd-ae8e-2e3fc290d6fd" />
